### PR TITLE
[ENH] DeepAR and  NHiTS and refinements for pytorch-forecasting interface

### PR DESCRIPTION
**Pull request by Xinyu-Wu-0000**
Add two additional forecaster to #6228.

About [`max_prediction_length=1` issue](https://github.com/jdb78/pytorch-forecasting/issues/1571):

NHiTS model in pytorch forecasting has a bug with heuristicly generating `pooling_sizes` and `downsample_frequencies` when `max_prediction_length=1`. Therefor  `pooling_sizes` and `downsample_frequencies` with values all 1 will be set manually if `max_prediction_length=1` for a workaround. A [PR #1577](https://github.com/jdb78/pytorch-forecasting/pull/1577) was opened in upstream repository to replace zeros in heuristicly generated `pooling_sizes` and `downsample_frequencies` with 1.

### CODE REVIEW

**Comment by Xinyu-Wu-0000** [comment]
@fkiraly Am I doing it right with a PR merging into main branch? It's recognizing all commits from #6228 as changes too.

**Comment by fkiraly** [comment]
Yes - the Pr description could say that #6228 should be merged first, but all in all this is the right pattern, i.e., branch off and open new PR. Once #6228 is merge, you should merge main into this.

**Comment by Xinyu-Wu-0000** [comment]
The notebook for global forecast is added in this PR, toward issue https://github.com/sktime/sktime/issues/6575.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Great! FYI @benHeid, have you seen and reviewed this?

I mainly have comments about the notebook.

* could you separate this into another PR? The notebook is great, but there is some iteration needed regarding location and integration with the other notebooks. I think we can merge the forecasters already and don't need to delay.

Some minor comments on the notebook content:
* there are a lot of printouts that confuse the reader. These should be silenced so you can focus on the didactic content.
* the markdown text is nice, I would just format it so the lines are not too long, and I would also use shorter telegram style, like on ppt slides

Regarding location, I would actually add the content to the notebook 01c, which has some content already. There is some minor confusion in the notebook about terminology, the notebook also uses the term "global" but in a different way.
* in the M5 paper, what the notebook does is called "cross-learning"
* the "global" in current 01d is more of a pre-training on other instances

In any case, we need to disambiguate terminology and perhaps adopt clearer distinctions here. @benHeid, what do you suggest on how we handle the terminology clash between 01c and 01d? And, should this go in the same notebook, so the "multiple instances"  cases can be explained easily?

**Comment by fkiraly** [comment]
Also FYI @shlok191 since you are looking at M5/M6 - do you have some further clarification or recommendation, terminology-wise?

**Comment by shlok191** [comment]
@fkiraly, @Xinyu-Wu-0000. Yes, I have some minor input! In the `01d_forecasting_global_forecast.ipynb` notebook and the `01c_forecasting_hierarchical_global.ipynb` notebooks, "global learning" is referred to as a term. I do agree that instead of using "global learning", we can instead use "pretrained" and "cross-learning" as replacements.

Here is how I'd differentiate them:
- To the best of my knowledge, pretrained models do not require for the training dataset time series to be correlated to each other

- Referencing this [paper on the M5 competition](https://www.sciencedirect.com/science/article/pii/S0169207021001874), I believe "cross-learning" is the term utilized for training models on time series which have a strong correlation. I believe M5 was referenced in 01c, so it might be good to use "cross-learning" there instead of "global learning"!

File: sktime/forecasting/pytorchforecasting.py [hunk 1]
```diff
@@ -30,6 +30,123 @@ class PytorchForecastingTFT(_PytorchForecastingAdapter):
...
+    ... )
+    >>> # define the model
+    >>> model = PytorchForecastingTFT(
+    ...     trainer_params={
+    ...         "max_epochs": 5,  # for quick test
+    ...         "limit_train_batches": 10,  # for quick test
+    ...     },
+    ... )
+    >>> # fit and predict
+    >>> model.fit(y=y_train, X=X_train, fh=fh)
```
Comment (benHeid) [inline sktime/forecasting/pytorchforecasting.py:71]: Do we want to add this lengthy output into the example? I would propose to not add such an output and perhaps skip this line using # doctest skip

File: sktime/forecasting/pytorchforecasting.py [hunk 2]
```diff
@@ -222,6 +339,103 @@ class PytorchForecastingNBeats(_PytorchForecastingAdapter):
...
+    >>> # define the model
+    >>> model = PytorchForecastingNBeats(
+    ...     trainer_params={
+    ...         "max_epochs": 5,  # for quick test
+    ...         "limit_train_batches": 10,  # for quick test
+    ...     },
+    ... )
+    >>> # fit and predict
+    >>> model.fit(y=y_train, fh=fh)
+    GPU available: True (cuda), used: True
```
Comment (benHeid) [inline sktime/forecasting/pytorchforecasting.py:379]: See comment above

File: sktime/forecasting/pytorchforecasting.py [hunk 3]
```diff
@@ -378,3 +592,575 @@ def get_test_params(cls, parameter_set="default"):
...
+    ... )
+    >>> # define the model
+    >>> model = PytorchForecastingDeepAR(
+    ...     trainer_params={
+    ...         "max_epochs": 5,  # for quick test
+    ...         "limit_train_batches": 10,  # for quick test
+    ...     },
+    ... )
+    >>> # fit and predict
+    >>> model.fit(y=y_train, X=X_train, fh=fh)
```
Comment (benHeid) [inline sktime/forecasting/pytorchforecasting.py:656]: See comment above

File: sktime/forecasting/pytorchforecasting.py [hunk 4]
```diff
@@ -378,3 +592,575 @@ def get_test_params(cls, parameter_set="default"):
...
+        Returns
+        -------
+        params : dict or list of dict, default = {}
+            Parameters to create testing instances of the class
+            Each dict are parameters to construct an "interesting" test instance, i.e.,
+            ``MyClass(**params)`` or ``MyClass(**params[i])`` creates a valid test
+            instance.
+            ``create_test_instance`` uses the first (or only) dictionary in ``params``
+        """
+        del parameter_set  # to avoid being detected as unused by ``vulture`` etc.
```
Comment (benHeid) [inline sktime/forecasting/pytorchforecasting.py:1097]: I would propose to remove this line. Since we do not do this in other tests.

File: sktime/forecasting/pytorchforecasting.py [hunk 5]
```diff
@@ -378,3 +592,575 @@ def get_test_params(cls, parameter_set="default"):
...
+                        "limit_train_batches": 10,  # for quick test
+                    },
+                    "dataset_params": {
+                        "max_encoder_length": 3,
+                    },
+                    "random_log_path": True,  # fix multiprocess file access error in CI
+                },
+                {
+                    "trainer_params": {
+                        "max_epochs": 3,  # for quick test
```
Comment (benHeid) [inline sktime/forecasting/pytorchforecasting.py:1115]: set it to one to save runtime

**Review by benHeid — CHANGES_REQUESTED** [review]
I focoussed on the code of the forecasters since @fkiraly focussed on the notebook. 

There are a few questions regarding the documentation

File: sktime/forecasting/pytorchforecasting.py (hunk 1 above)
Comment (Xinyu-Wu-0000) [reply sktime/forecasting/pytorchforecasting.py]: lengthy output is removed and # doctest skip is added

File: sktime/forecasting/pytorchforecasting.py (hunk 2 above)
Comment (Xinyu-Wu-0000) [reply sktime/forecasting/pytorchforecasting.py]: lengthy output is removed and # doctest skip is added

File: sktime/forecasting/pytorchforecasting.py (hunk 5 above)
Comment (Xinyu-Wu-0000) [reply sktime/forecasting/pytorchforecasting.py]: done

**Comment by Xinyu-Wu-0000** [comment]
FYI @fkiraly @shlok191 The notebook is moved to https://github.com/sktime/sktime/pull/6699.

File: sktime/forecasting/base/adapters/_pytorchforecasting.py [hunk 6]
```diff
@@ -39,7 +39,11 @@ class _PytorchForecastingAdapter(_BaseGlobalForecaster):
         parameters to be passed for `TimeSeriesDataSet.to_dataloader()`
         by default {"train": False}
     model_path: string (default=None)
-        try to load a existing model without fitting.
+        try to load a existing model without fitting. Calling the fit function is
```
Comment (benHeid) [inline sktime/forecasting/base/adapters/_pytorchforecasting.py:42]: I wonder, if we should also allow to pass an initialized pytorch-forecasting model and not only a path where it is stored. 

Any opinions? Could also be solved via a different PR.

File: sktime/forecasting/base/adapters/_pytorchforecasting.py [hunk 7]
```diff
@@ -314,6 +327,11 @@ def _predict(
                 else None
             ),
         )
+        try:
```
Comment (benHeid) [inline sktime/forecasting/base/adapters/_pytorchforecasting.py:330]: Can you extract this check into a private method and use it also in line 310-318? 

Is there a reason why you haven't used `torch.manual_seed` here?

File: sktime/forecasting/pytorchforecasting.py [hunk 8]
```diff
@@ -222,6 +295,79 @@ class PytorchForecastingNBeats(_PytorchForecastingAdapter):
...
+                2000-02-19  5.372994
+    ...                          ...
+    h0_4 h1_199 2000-02-15  5.005799
+                2000-02-16  4.998720
+                2000-02-17  5.031197
+                2000-02-18  5.081184
+                2000-02-19  5.113482
+
+    [500 rows x 1 columns]
+    >>>
```
Comment (benHeid) [inline sktime/forecasting/pytorchforecasting.py:370]: why are these three `>>>` here. If not required by dostring, please remove them. This applies to all of the examples.

File: sktime/forecasting/pytorchforecasting.py [hunk 9]
```diff
@@ -378,3 +522,520 @@ def get_test_params(cls, parameter_set="default"):
...
+                2000-02-19  4.808967
+    ...                          ...
+    h0_4 h1_199 2000-02-15  5.150748
+                2000-02-16  5.230827
+                2000-02-17  5.123736
+                2000-02-18  5.139505
+                2000-02-19  5.121511
+
+    [500 rows x 1 columns]
+    >>>
```
Comment (benHeid) [inline sktime/forecasting/pytorchforecasting.py:622]: why are these three `>>>` here. If not required by dostring, please remove them. This applies to all of the examples.

File: sktime/forecasting/pytorchforecasting.py [hunk 10]
```diff
@@ -378,3 +522,520 @@ def get_test_params(cls, parameter_set="default"):
...
+                2000-02-19  5.760923
+    ...                          ...
+    h0_4 h1_199 2000-02-15  5.376267
+                2000-02-16  5.227071
+                2000-02-17  5.070744
+                2000-02-18  5.249713
+                2000-02-19  5.047630
+
+    [500 rows x 1 columns]
+    >>>
```
Comment (benHeid) [inline sktime/forecasting/pytorchforecasting.py:882]: why are these three `>>>` here. If not required by dostring, please remove them. This applies to all of the examples.

**Review by benHeid — CHANGES_REQUESTED** [review]
Thank you, I have some additional small remarks (sorry that I haven't seen them before).

File: sktime/forecasting/base/adapters/_pytorchforecasting.py (hunk 6 above)
Comment (Xinyu-Wu-0000) [reply sktime/forecasting/base/adapters/_pytorchforecasting.py:42]: > I wonder, if we should also allow to pass an initialized pytorch-forecasting model and not only a path where it is stored.

Yeah, it could be done by passing a model and assigning it to `self.best_model`.

File: sktime/forecasting/base/adapters/_pytorchforecasting.py (hunk 7 above)
Comment (Xinyu-Wu-0000) [reply sktime/forecasting/base/adapters/_pytorchforecasting.py:330]: I am not sure how to get the original seed, and state is more precise than a seed I think.

File: sktime/forecasting/base/adapters/_pytorchforecasting.py (hunk 6 above)
Comment (Xinyu-Wu-0000) [reply sktime/forecasting/base/adapters/_pytorchforecasting.py:42]: We can make it capable of fine tuning actully, loading from a existing model and than tuning it. I am not sure whether it's consistant with the fine tuning api design though, maybe we can wait until the fine tuning dessign is finished.

File: sktime/forecasting/pytorchforecasting.py (hunk 10 above)
Comment (Xinyu-Wu-0000) [reply sktime/forecasting/pytorchforecasting.py]: removed

File: sktime/forecasting/pytorchforecasting.py (hunk 9 above)
Comment (Xinyu-Wu-0000) [reply sktime/forecasting/pytorchforecasting.py]: removed

File: sktime/forecasting/pytorchforecasting.py (hunk 8 above)
Comment (Xinyu-Wu-0000) [reply sktime/forecasting/pytorchforecasting.py]: removed

File: sktime/forecasting/base/adapters/_pytorchforecasting.py (hunk 6 above)
Comment (benHeid) [reply sktime/forecasting/base/adapters/_pytorchforecasting.py:42]: I am not sure, if these models can be easily used within fine-tuning, since for fine-tuning we need direct access to the neural network so I am unsure if this is possible with pytorch-forecasting easily.

I think it would be cool, if you can create an issue for adding support to pass a model directly.

**Review by benHeid — DISMISSED** [review]

File: sktime/forecasting/base/adapters/_pytorchforecasting.py (hunk 6 above)
Comment (Xinyu-Wu-0000) [reply sktime/forecasting/base/adapters/_pytorchforecasting.py:42]: issue is created: https://github.com/sktime/sktime/issues/6736

**Comment by Xinyu-Wu-0000** [comment]
@benHeid after merging main into this PR, `Undefined name Dict` and  `Undefined name List` errors occurred, so I have to push a [fix commit](https://github.com/sktime/sktime/pull/6551/commits/ba4b6faabe05fcffcf4f9bdc4e36265dc6f58b45).

**Comment by benHeid** [comment]
I suppose you need to execute ruff locally.

**Comment by Xinyu-Wu-0000** [comment]
> I suppose you need to execute ruff locally.

Yeah, I updated my pre-commit installation and env setup.

**Comment by benHeid** [comment]
@Xinyu-Wu-0000 do you have any ideas, what is happening with the failing test? There are no useful logs as far as I see. And restarting was not helpful..

File: sktime/forecasting/pytorchforecasting.py [hunk 11]
```diff
@@ -17,20 +17,95 @@ class PytorchForecastingTFT(_PytorchForecastingAdapter):
...
+    >>> from sktime.utils._testing.hierarchical import _make_hierarchical
+    >>> # generate random data
+    >>> data = _make_hierarchical(
+    ...     hierarchy_levels=(5, 200), max_timepoints=50, min_timepoints=50, n_columns=3
+    ... )
+    >>> # define forecast horizon
+    >>> max_prediction_length = 5
+    >>> fh = ForecastingHorizon(range(1, max_prediction_length + 1), is_relative=True)
+    >>> # split X, y data for train and test
+    >>> l1 = data.index.get_level_values(1).map(lambda x: int(x[3:]))
```
Comment (benHeid) [inline sktime/forecasting/pytorchforecasting.py:55]: Can you check if such a split is possible to implement with the InstanceSplitter? This would make the docs much easier to read. If not possible, I think it would be cool to add a detailed comment, since this is visible on the documentation page. Thus, it should be easy to understand.

Comment (benHeid) [reply sktime/forecasting/pytorchforecasting.py:55]: This would apply to all PyTorch forecaster I assume

File: sktime/forecasting/tests/test_pytorchforecasting.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    # generate data
+    data_length = 100
+    data = _make_hierarchical(
+        (5, 100),
+        n_columns=2,
+        max_timepoints=data_length,
+        min_timepoints=data_length,
+    )
+    l1 = data.index.get_level_values(1).map(lambda x: int(x[3:]))
+    X_train = data.loc[l1 < 90, "c0"].to_frame()
```
Comment (benHeid) [inline sktime/forecasting/tests/test_pytorchforecasting.py:58]: Also here please check if this is the same as InstanceSplitter is doing.

**Review by benHeid — COMMENTED** [review]
Just, one small thing regarding the splits. Please check if the InstanceSplitter could be used here. 
I think it would also be useful, to implement something similar as the split_train_test method. What do you think? If you agree, it would be cool, if you would create an issue :)

**Comment by Xinyu-Wu-0000** [comment]
> Just, one small thing regarding the splits. Please check if the InstanceSplitter could be used here.
> I think it would also be useful, to implement something similar as the split_train_test method. What do you think? If you agree, it would be cool, if you would create an issue :)

`sklearn.model_selection.train_test_split` can do it. `train_test_split` can only split data to train and test dataset without any overlap, however what the tests do is to have train and test dataset with overlap to achieve better coverage. Therefor two `train_test_split` have to be used in tests to achieve the same splitting.

Since we can just borrow `train_test_split` function from `sklearn`, I will simply add a commit in this PR.

> @Xinyu-Wu-0000 do you have any ideas, what is happening with the failing test? There are no useful logs as far as I see. And restarting was not helpful..

I tried to run test_full in a docker container locally, could not replicate. Hope a merge of main will fix it.

File: sktime/forecasting/pytorchforecasting.py (hunk 11 above)
Comment (Xinyu-Wu-0000) [reply sktime/forecasting/pytorchforecasting.py]: replaced with `sklearn.model_selection.train_test_split`

File: sktime/forecasting/tests/test_pytorchforecasting.py (hunk 12 above)
Comment (Xinyu-Wu-0000) [reply sktime/forecasting/tests/test_pytorchforecasting.py]: replaced with `sklearn.model_selection.train_test_split`

**Comment by benHeid** [comment]
@Xinyu-Wu-0000 do you see any logs for the failing test?

**Comment by benHeid** [comment]
image
It seems that there is not enough space left on the runner... 

when letting the overview open during the execution, I saw that also some of the tests failed:

image

However, I am unable to figure out which one since the memory issue occurred again..

**Comment by Xinyu-Wu-0000** [comment]
@benHeid all failing tests are resolved. Thanks for the help.

**Comment by Xinyu-Wu-0000** [comment]
@benHeid The not enough space issue is also happening in neuralforecaset PR and huggingface PR. Should we merge this  soon?

**Review by benHeid — DISMISSED** [review]
Looks good to me. 

@fkiraly do you have additional comments?

**Review by fkiraly — DISMISSED** [review]
Looks good, no additional comments.

One note, the tag `python_dependencies` should be `pytorch-forecasting`, not `pytorch_forecastrng`, because that´s the pypi name

**Comment by Xinyu-Wu-0000** [comment]
@benHeid @fkiraly  `_check_soft_dependencies` related problems of DeepAR and NHiTS are fixed directly in this PR.

**Review by benHeid — APPROVED** [review]
